community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/DegreeCounter.java

```diff
@@ -0,0 +1,104 @@
+    {
+        if ( relationshipId == Record.NO_NEXT_RELATIONSHIP.intValue() )
+        {
+            return 0;
+        }
+        cursor.next( relationshipId, relationshipRecord, FORCE );
+        if ( relationshipRecord.getFirstNode() == nodeRecord.getId() )
+        {
+            return relationshipRecord.getFirstPrevRel();
+        }
+        if ( relationshipRecord.getSecondNode() == nodeRecord.getId() )
+        {
+            return relationshipRecord.getSecondPrevRel();
+        }
+        throw new InvalidRecordException( "Node " + nodeRecord.getId() + " neither start nor end node of " + relationshipRecord );
+    }
+
+    default int countRelationshipsInGroup( long groupId, Direction direction, Integer type, NodeRecord nodeRecord,
+            RelationshipRecord relationshipRecord, RelationshipGroupRecord groupRecord, RecordCursors cursors )
+    {
+        int count = 0;
+        while ( groupId != Record.NO_NEXT_RELATIONSHIP.intValue() )
+        {
+            boolean groupRecordInUse = cursors.relationshipGroup().next( groupId, groupRecord, FORCE );
+            if ( groupRecordInUse && ( type == null || groupRecord.getType() == type ) )
+            {
+                count += nodeDegreeByDirection( direction, nodeRecord, relationshipRecord, groupRecord, cursors );
+                if ( type != null )
+                {
+                    // we have read the only type we were interested on, so break the look
+                    break;
+                }
+            }
+            groupId = groupRecord.getNext();
+        }
+        return count;
+    }
+
+    default long nodeDegreeByDirection( Direction direction, NodeRecord nodeRecord,
+            RelationshipRecord relationshipRecord, RelationshipGroupRecord groupRecord, RecordCursors cursors )
+    {
+        RecordCursor<RelationshipRecord> cursor = cursors.relationship();
+        long loopCount = countByFirstPrevPointer( groupRecord.getFirstLoop(), cursor, nodeRecord, relationshipRecord );
+        switch ( direction )
+        {
+        case OUTGOING:
+        {
+            long firstOut = groupRecord.getFirstOut();
+            return countByFirstPrevPointer( firstOut, cursor, nodeRecord, relationshipRecord ) + loopCount;
+        }
+        case INCOMING:
+        {
+            long firstIn = groupRecord.getFirstIn();
+            return countByFirstPrevPointer( firstIn, cursor, nodeRecord, relationshipRecord ) + loopCount;
+        }
+        case BOTH:
+        {
+            long firstOut = groupRecord.getFirstOut();
+            long firstIn = groupRecord.getFirstIn();
+            return countByFirstPrevPointer( firstOut, cursor, nodeRecord, relationshipRecord ) +
+                    countByFirstPrevPointer( firstIn, cursor, nodeRecord, relationshipRecord ) + loopCount;
+        }
+        default:
+            throw new IllegalArgumentException( direction.name() );
+        }
+    }
```

Comment:
Maybe you want these methods to be static instead of default? They don't really rely on any methods in this or super interfaces. Default methods are primarily for adding methods to existing interfaces in a backwards compatible way. They're also harder for the JIT to optimise, because of their complicated resolution logic.